### ISSUE TITLE
Concurrency issue porting to a new platform

### ISSUE THREAD

**Issue by LeanderBB** [microsoft/mimalloc#421]
Hi,

I'm currently trying to port this project to a platform which I sadly can't discuss in the open. I can say that it runs a x86_64 CPU though. I was wondering whether someone could provide me some guidance into what I should double check to ensure the porting is successful? I may have overlooked some, hopefully, trivial detail.

I've currently replaced all `_mi_os_*` and the random generator calls with the platform's equivalent API. I can successfully run the mimalloc-test-api project. 

That being said, once I  start the mimalloc-test-stress project, things get a bit hairy. When running in Debug with MI_DEBUG_FULL mode I get the following list of asserts (and callstacks) at random:
```
mimalloc: assertion failed: at "D:\mimalloc\src\page.c":112, _mi_page_is_valid
  assertion: "!_mi_process_is_initialized || segment->thread_id == mi_page_heap(page)->thread_id || segment->thread_id==0" !_mi_page_is_valid(mi_page_t* page) Line 113	C++
!mi_page_fresh_alloc(mi_heap_t* heap, mi_page_queue_t* pq, size_t block_size) Line 256 + 9 bytes	C++
!mi_page_fresh(mi_heap_t* heap, mi_page_queue_t* pq) Line 263 + 21 bytes	C++
!mi_page_queue_find_free_ex(mi_heap_t* heap, mi_page_queue_t* pq, bool first_try) Line 687 + 13 bytes	C++
!mi_find_free_page(mi_heap_t* heap, size_t size) Line 724 + 18 bytes	C++
!mi_find_page(mi_heap_t* heap, size_t size) Line 804 + 13 bytes	C++
!_mi_malloc_generic(mi_heap_t* heap, size_t size) Line 829 + 13 bytes	C++
!_mi_page_malloc(mi_heap_t* heap, mi_page_t* page, size_t size) Line 28 + 13 bytes	C++
!mi_heap_malloc_small(mi_heap_t* heap, size_t size) Line 79 + 27 bytes	C++
!mi_heap_malloc(mi_heap_t* heap, size_t size) Line 97 + 13 bytes	C++
!_mi_heap_malloc_zero(mi_heap_t* heap, size_t size, bool zero) Line 147 + 13 bytes	C++
!mi_heap_zalloc(mi_heap_t* heap, size_t size) Line 155 + 18 bytes	C++
!mi_heap_calloc(mi_heap_t* heap, size_t count, size_t size) Line 593 + 13 bytes	C++
!mi_calloc(size_t count, size_t size) Line 597 + 31 bytes	C++
!alloc_items(size_t items, random_t r) Line 101 + 14 bytes	C++
!stress(intptr_t tid) Line 150 + 44 bytes	C++
  

mimalloc: assertion failed: at "D:\mimalloc\src\random.c":145, _mi_random_next
  assertion: "mi_random_is_initialized(ctx)"
!_mi_assert_fail(const char* assertion, const char* fname, unsigned int line, const char* func) Line 344 + 5 bytes	C++
!_mi_random_next(mi_random_ctx_t* ctx) Line 149	C++
!_mi_heap_random_next(mi_heap_t* heap) Line 211 + 18 bytes	C++
!mi_page_init(mi_heap_t* heap, mi_page_t* page, size_t block_size, mi_tld_t* tld) Line 620 + 9 bytes	C++
!mi_page_fresh_alloc(mi_heap_t* heap, mi_page_queue_t* pq, size_t block_size) Line 254	C++
!mi_page_fresh(mi_heap_t* heap, mi_page_queue_t* pq) Line 263 + 21 bytes	C++
!mi_page_queue_find_free_ex(mi_heap_t* heap, mi_page_queue_t* pq, bool first_try) Line 687 + 13 bytes	C++
!mi_find_free_page(mi_heap_t* heap, size_t size) Line 724 + 18 bytes	C++
!mi_find_page(mi_heap_t* heap, size_t size) Line 804 + 13 bytes	C++
!_mi_malloc_generic(mi_heap_t* heap, size_t size) Line 829 + 13 bytes	C++
!_mi_page_malloc(mi_heap_t* heap, mi_page_t* page, size_t size) Line 28 + 13 bytes	C++
!mi_heap_malloc_small(mi_heap_t* heap, size_t size) Line 79 + 27 bytes	C++
!mi_heap_malloc(mi_heap_t* heap, size_t size) Line 97 + 13 bytes	C++
!_mi_heap_malloc_zero(mi_heap_t* heap, size_t size, bool zero) Line 147 + 13 bytes	C++
!mi_heap_zalloc(mi_heap_t* heap, size_t size) Line 155 + 18 bytes	C++
!mi_heap_calloc(mi_heap_t* heap, size_t count, size_t size) Line 593 + 13 bytes	C++
!mi_calloc(size_t count, size_t size) Line 597 + 31 bytes	C++
!alloc_items(size_t items, random_t r) Line 101 + 14 bytes	C++
!stress(intptr_t tid) Line 150 + 44 bytes	C++

  
mimalloc: assertion failed: at "D:\mimalloc\src\segment.c":163, mi_segment_is_valid
  assertion: "nfree + segment->used == segment->capacity"
!_mi_assert_fail(const char* assertion, const char* fname, unsigned int line, const char* func) Line 344 + 5 bytes	C++
!mi_segment_is_valid(const mi_segment_t* segment, mi_segments_tld_t* tld) Line 165	C++
!_mi_segment_page_alloc(mi_heap_t* heap, size_t block_size, mi_segments_tld_t* tld, mi_os_tld_t* os_tld) Line 1365 + 32 bytes	C++
!mi_page_fresh_alloc(mi_heap_t* heap, mi_page_queue_t* pq, size_t block_size) Line 246 + 43 bytes	C++
!mi_page_fresh(mi_heap_t* heap, mi_page_queue_t* pq) Line 263 + 21 bytes	C++
!mi_page_queue_find_free_ex(mi_heap_t* heap, mi_page_queue_t* pq, bool first_try) Line 687 + 13 bytes	C++
!mi_find_free_page(mi_heap_t* heap, size_t size) Line 724 + 18 bytes	C++
!mi_find_page(mi_heap_t* heap, size_t size) Line 804 + 13 bytes	C++
!_mi_malloc_generic(mi_heap_t* heap, size_t size) Line 829 + 13 bytes	C++
!mi_heap_malloc(mi_heap_t* heap, size_t size) Line 102 + 22 bytes	C++
!_mi_heap_malloc_zero(mi_heap_t* heap, size_t size, bool zero) Line 147 + 13 bytes	C++
!mi_heap_zalloc(mi_heap_t* heap, size_t size) Line 155 + 18 bytes	C++
!mi_heap_calloc(mi_heap_t* heap, size_t count, size_t size) Line 593 + 13 bytes	C++
!mi_calloc(size_t count, size_t size) Line 597 + 31 bytes	C++
!stress(intptr_t tid) Line 135 + 12 bytes	C++
  

mimalloc: assertion failed: at "D:\mimalloc\src\alloc.c":71, mi_heap_malloc_small
  assertion: "heap->thread_id == 0 || heap->thread_id == _mi_thread_id()"
!_mi_assert_fail(const char* assertion, const char* fname, unsigned int line, const char* func) Line 344 + 5 bytes	C++
!mi_heap_malloc_small(mi_heap_t* heap, size_t size) Line 71 + 87 bytes	C++
!mi_heap_malloc(mi_heap_t* heap, size_t size) Line 97 + 13 bytes	C++
!_mi_heap_malloc_zero(mi_heap_t* heap, size_t size, bool zero) Line 147 + 13 bytes	C++
!mi_heap_zalloc(mi_heap_t* heap, size_t size) Line 155 + 18 bytes	C++
!mi_heap_calloc(mi_heap_t* heap, size_t count, size_t size) Line 593 + 13 bytes	C++
!mi_calloc(size_t count, size_t size) Line 597 + 31 bytes	C++
!alloc_items(size_t items, random_t r) Line 101 + 14 bytes	C++
!stress(intptr_t tid) Line 150 + 44 bytes	C++
```
And a consistent, albeit random, memory crashes in `mi_stat_update`.
```
!mi_stat_update(mi_stat_count_t* stat, int64_t amount) Line 43 + 8 bytes	C++
!_mi_stat_increase(mi_stat_count_t* stat, size_t amount) Line 67	C++
!mi_page_fresh_alloc(mi_heap_t* heap, mi_page_queue_t* pq, size_t block_size) Line 255	C++
!mi_page_fresh(mi_heap_t* heap, mi_page_queue_t* pq) Line 263 + 21 bytes	C++
!mi_page_queue_find_free_ex(mi_heap_t* heap, mi_page_queue_t* pq, bool first_try) Line 687 + 13 bytes	C++
!mi_find_free_page(mi_heap_t* heap, size_t size) Line 724 + 18 bytes	C++
!mi_find_page(mi_heap_t* heap, size_t size) Line 804 + 13 bytes	C++
!_mi_malloc_generic(mi_heap_t* heap, size_t size) Line 829 + 13 bytes	C++
!_mi_page_malloc(mi_heap_t* heap, mi_page_t* page, size_t size) Line 28 + 13 bytes	C++
!mi_heap_malloc_small(mi_heap_t* heap, size_t size) Line 79 + 27 bytes	C++
!mi_heap_malloc(mi_heap_t* heap, size_t size) Line 97 + 13 bytes	C++
!_mi_heap_malloc_zero(mi_heap_t* heap, size_t size, bool zero) Line 147 + 13 bytes	C++
!mi_heap_zalloc(mi_heap_t* heap, size_t size) Line 155 + 18 bytes	C++
!mi_heap_calloc(mi_heap_t* heap, size_t count, size_t size) Line 593 + 13 bytes	C++
!mi_calloc(size_t count, size_t size) Line 597 + 31 bytes	C++
!alloc_items(size_t items, random_t r) Line 101 + 14 bytes	C++
!stress(intptr_t tid) Line 150 + 44 bytes	C++
```

There's clearly something not right when allocating from different threads. Most of the asserts above trigger as soon as there are 2 stress threads active.

Here's what I've tried so far:

- Use standard C approach of getting threadIds instead of inline asm, but the above asserts are still triggering.
- Tried different approaches to get thread local data, they all work fine in standalone examples. But, seeing as most of these problem end up with ThreadIDs not matching and the fact that the heap object is retrieved from a thread local variable, I am not a 100% sure. 
- Tried enabling MI_SECURE and MI_SHOW_ERRORS
- I'm compiling the code as CXX (MI_USE_CXX) since the compiler for that platform does not have a C11 compatible implementation and am relying on C++ atomics. I'm assuming that the use of C++ shouldn't be an issue since there this is being used for Windows. I haven't discarded the possibility they may  be an issue with atomic lib on this platform, but that seems a bit unlikely.
- Sadly TSAN'/UBSAN aren't available on this platform :(
- The Platform's APIs work fine in a multi-threaded scenario. 
- Disabled MI_OVERRIDE

Any insights, ideas, pointers on what I should pay attention to/check/ensure in order to track down the above issue, if possible? Is there some guarantee that the platform needs to have for this code to work properly?

I totally understand if no help can be provided due to the nature of the platform.

Thanks in advance for your time.

Kind Regards,
Leander

EDIT: I forgot to add, in case it is relevant, I'm using the static library config.

EDIT2: I'm using the master branch at commit 076f815

**Comment by LeanderBB** [comment]
I forgot to add, in case it is relevant, I'm using the static library config.

**Comment by daanx** [comment]
Ah very strange errors; I think indeed something is off with either the thread id's and/or initialization. 

- You already tried the portable C for the thread id ... so that should just work for thread id's I think ; note that the inline assemble on x86_64 uses the first entry pointed to by the GS or FS register which depends on the platform (mach/macos use GS while Linux/BSD use FS); you may need to experiment which one is used on your secret OS.

- Also, you use CXX which is good: mimalloc needs to initialize at process startup; see "init.c" at the end where I use a C++ global class to initialize before "main" runs. Check if that is actually working on your platform. Or perhaps call "mi_process_init" manually at the start of "main" in the stress test to rule that out as a possibility.

- Maybe the TLS implementation on your secret OS is not working well? Perhaps try passing `-DMI_LOCAL_DYNAMIC_TLS=ON` to cmake ?

Let me know how it goes

**Comment by LeanderBB** [comment]
- I'll definitely look into the ASM option once I've fixed the current issue!
- I've verified that `mi_process_init` gets called through the static CXX initializer.
- The `MI_LOCAL_DYNAMIC_TLS` option has no effect on the results.

I'm starting to think I may not be using the local APIs  properly. I will get back to you. Thank for the pointers.

**Comment by LeanderBB** [comment]
Ok, so I found the problem. There's nothing wrong with mimalloc, it was user error all along!

Due to some severe tunnel vision, I failed to observe some API misuse that was causing multiple instances of virtual memory to map to same physical memory block. After correcting that mistake, all test pass with flying colors :).

I thank you again for your time as it at least helped me to rule out some other issues :+1:
